mlir/lib/Analysis/Presburger/Matrix.cpp

```diff
@@ -548,4 +548,29 @@ Fraction FracMatrix::determinant(FracMatrix *inverse) const {
     determinant *= m.at(i, i);
 
   return determinant;
+}
+
+FracMatrix FracMatrix::gramSchmidt() const {
+  bool linIndep =
+      (nRows < nColumns) || (nRows == nColumns && determinant(nullptr) != 0);
+  assert(linIndep && "the vectors must be linearly independent!");
+
+  // Create a copy of the argument to store
+  // the orthogonalised version.
+  FracMatrix orth(*this);
+
+  // For each vector (row) in the matrix,
+  // subtract its unit projection along
+  // each of the previous vectors.
+  // This ensures that it has no component
+  // in the direction of any of the
+  // previous vectors.
+  for (unsigned i = 1; i < getNumRows(); i++) {
```

Comment:
for (i = 1, e = getNumRows(); i < e; ++i)

Reply:
https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

Reply:
Please apply to all loops.